pkg/lib/bundle/validate.go

```diff
@@ -0,0 +1,142 @@
+package bundle
+
+import (
+	"fmt"
+	"io/ioutil"
+	"path/filepath"
+	"strings"
+
+	"github.com/operator-framework/operator-registry/pkg/containertools"
+
+	log "github.com/sirupsen/logrus"
+	"gopkg.in/yaml.v2"
+)
+
+// imageValidator is a struct implementation of the Indexer interface
+type imageValidator struct {
+	imageReader containertools.ImageReader
+	logger      *log.Entry
+}
+
+// PullBundleImage shells out to a container tool and pulls a given image tag
+// Then it unpacks the image layer filesystem contents and pushes the contents
+// to a specified directory for further validation
+func (i imageValidator) PullBundleImage(imageTag, directory string) error {
+	i.logger.Debug("Pulling and unpacking container image")
+
+	return i.imageReader.GetImageData(imageTag, directory)
+}
+
+// ValidateBundle takes a directory containing the contents of a bundle and validates
+// the format of that bundle for correctness
+func (i imageValidator) ValidateBundle(directory string) error {
+	var manifestsFound, metadataFound bool
+	var annotationsDir, manifestsDir string
+	var annotationErrors []error
+	var formatErrors []error
+
+	items, _ := ioutil.ReadDir(directory)
+	for _, item := range items {
+		if item.IsDir() {
+			switch s := item.Name(); s {
+			case strings.TrimSuffix(ManifestsDir, "/"):
+				i.logger.Debug("Found manifests directory")
+				manifestsFound = true
+				manifestsDir = filepath.Join(directory, ManifestsDir)
+			case strings.TrimSuffix(MetadataDir, "/"):
+				i.logger.Debug("Found metadata directory")
+				metadataFound = true
+				annotationsDir = filepath.Join(directory, MetadataDir)
+			}
+		}
+	}
+
+	if manifestsFound == false {
+		formatErrors = append(formatErrors, fmt.Errorf("Unable to locate manifests directory"))
+	}
+	if metadataFound == false {
+		formatErrors = append(formatErrors, fmt.Errorf("Unable to locate metadata directory"))
+	}
+
+	// Break here if we can't even find the files
+	if len(formatErrors) > 0 {
+		return NewValidationError(annotationErrors, formatErrors)
+	}
+
+	i.logger.Debug("Getting mediaType info from manifests directory")
+	mediaType, err := GetMediaType(manifestsDir)
+	if err != nil {
+		formatErrors = append(formatErrors, err)
+	}
+
+	// Validate annotations.yaml
+	annotationsFile, err := ioutil.ReadFile(filepath.Join(annotationsDir, AnnotationsFile))
+	if err != nil {
+		fmtErr := fmt.Errorf("Unable to read annotations.yaml file: %s", err.Error())
+		formatErrors = append(formatErrors, fmtErr)
+	}
```

Comment:
This method should return here since `annotationsFile` is used below, unless the intent is to run through the method body with a potentially empty `[]byte` and append resulting errors below.